### PR TITLE
many: derive implicit slots from interface meta-data

### PR DESCRIPTION
**Pull request by zyga**
This patch moves the declaration that a given interface should have an
implicit slot on the core snap from snap/implicit.go into each
particular interface definition file. This brings us closer to having
all the relevant code for a given interface be entirely defined in the
interface file itself.

The fuse-support interface had some extra complexity that was now merged
into the interface definition (it is not implicitly added on ubuntu
14.04). The interface code had to be changed from commonInterface to a
new dedicated type.

The code that adds interfaces is now moved to overlord/ifacecstate and
made private (it is only relevant there and it avoids cyclic imports).

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: overlord/ifacestate/ifacestate.go [hunk 1]
```diff
@@ -170,8 +170,8 @@ func Disconnect(st *state.State, plugSnap, plugName, slotSnap, slotName string)
 
 // CheckInterfaces checks whether plugs and slots of snap are allowed for installation.
 func CheckInterfaces(st *state.State, snapInfo *snap.Info) error {
-	// XXX: AddImplicitSlots is really a brittle interface
-	snap.AddImplicitSlots(snapInfo)
+	// XXX: addImplicitSlots is really a brittle interface
+	addImplicitSlots(snapInfo)
```
Comment (pedronis) [inline overlord/ifacestate/ifacestate.go]: can't we try to address the problem that is still brittle given that we are changing things ?

Comment (zyga) [inline overlord/ifacestate/ifacestate.go]: Yes but this is not something I wish to do in this branch. The goal right now is to reduce the number of files that need to be patched to create a new interface. Addressing the implicit slot mechanism can and should be done separately IMO. I think this branch does improve the situation slightly as it moves the code closer to the interface manager where ultimately all the interface decisions are being taken.

File: interfaces/builtin/fuse_support.go [hunk 2]
```diff
@@ -71,12 +81,52 @@ deny /etc/fuse.conf r,
...
+	return true
+}
+
 func init() {
-	registerIface(&commonInterface{
-		name:                  "fuse-support",
-		summary:               fuseSupportSummary,
-		connectedPlugAppArmor: fuseSupportConnectedPlugAppArmor,
-		connectedPlugSecComp:  fuseSupportConnectedPlugSecComp,
-		reservedForOS:         true,
```
Comment (mvo5) [inline interfaces/builtin/fuse_support.go]: Why is this not just adding something like:
```
registerIface(&commonInterface{
...
ImplicitOnClassic: !(release.ReleaseInfo.ID == "ubuntu" && release.ReleaseInfo.VersionID == "14.04"),
...
}
```
i.e. why did it have to be converted to a new type?

File: snap/implicit_test.go [hunk 3]
```diff
@@ -20,56 +20,9 @@
...
-	info, err := snap.InfoFromSnapYaml(osYaml)
-	c.Assert(err, IsNil)
-	snap.AddImplicitSlots(info)
-	c.Assert(info.Slots["unity7"].Interface, Equals, "unity7")
-	c.Assert(info.Slots["unity7"].Name, Equals, "unity7")
-	c.Assert(info.Slots["unity7"].Snap, Equals, info)
-	// Ensure that we have *some* implicit slots
-	c.Assert(len(info.Slots) > 10, Equals, true)
-}
+type implicitSuite struct{}
```
Comment (mvo5) [inline snap/implicit_test.go]: It seems this file is pretty empty now, why do we still have it?

File: interfaces/builtin/account_control.go [hunk 4]
```diff
@@ -72,6 +72,8 @@ func init() {
 		name:                  "account-control",
 		summary:               accountControlSummary,
 		description:           accountControlDescription,
+		implicitOnCore:        true,
```
Comment (mvo5) [inline interfaces/builtin/account_control.go]: Is there/will there ever be a case where we have `implicitOnCore: false, implicitOnClassic: true` ? The old code had `implicit` and `implicitOnlyOnClassic` only. Mostly wondering. Also wondering if: `implicit: All`, `implicit: CoreOnly`, `implicit: ClassicOnly` might look nicer (i.e. using bits).

Comment (zyga) [inline interfaces/builtin/account_control.go]: There are things that are implicit on classic, yes. I changed this from the previous code to make it less magic and more obvious as to what is added. I don't mind a single implicit bit-flag but I'd rather do that in a separate branch to allow this one to land and iterate.

File: snap/implicit_test.go (hunk 3 above)
Comment (zyga) [inline snap/implicit_test.go]: There are some things that we could test, I think. I can remove it or fill in the tests.

File: interfaces/builtin/fuse_support.go (hunk 2 above)
Comment (zyga) [inline interfaces/builtin/fuse_support.go]: Indeed, I think it could be a common interface.

File: interfaces/builtin/fuse_support.go [hunk 5]
```diff
@@ -72,11 +76,14 @@ deny /etc/fuse.conf r,
 `
 
 func init() {
+	// Ubuntu 14.04 does not support the fuse-support interface.
 	registerIface(&commonInterface{
 		name:                  "fuse-support",
 		summary:               fuseSupportSummary,
+		implicitOnCore:        true,
+		implicitOnClassic:     !(release.ReleaseInfo.ID == "ubuntu" && release.ReleaseInfo.VersionID == "14.04"),
```
Comment (jdstrand) [inline interfaces/builtin/fuse_support.go]: This looks like an unrelated change. I'm not saying it is wrong.... It seems that this sort of thing should be captured elsewhere. Eg, I can imagine Fedora or SUSE might have different interfaces. I mention this only because if going that direction, maybe it's best to not include this change in this PR.

File: overlord/ifacestate/implicit_test.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+	osYaml := []byte("name: ubuntu-core\ntype: os\n")
+	info, err := snap.InfoFromSnapYaml(osYaml)
+	c.Assert(err, IsNil)
+	ifacestate.AddImplicitSlots(info)
+	c.Assert(info.Slots["network"].Interface, Equals, "network")
+	c.Assert(info.Slots["network"].Name, Equals, "network")
+	c.Assert(info.Slots["network"].Snap, Equals, info)
+	// Ensure that we have *some* implicit slots
+	c.Assert(len(info.Slots) > 10, Equals, true)
```
Comment (jdstrand) [inline overlord/ifacestate/implicit_test.go]: Can we test that something only on classic is not in core?

File: interfaces/builtin/fuse_support.go (hunk 5 above)
Comment (zyga) [inline interfaces/builtin/fuse_support.go]: This is not an unrelated change. It used to be defined exactly this way in `implicit.go` before.

File: overlord/ifacestate/implicit_test.go (hunk 6 above)
Comment (zyga) [inline overlord/ifacestate/implicit_test.go]: Yes, sure :-)
